File: 0117_Populating Next Right Pointers in Each Node II/PopulatingNextRightPointersinEachNodeII.java
class Solution {
    public Node connect(Node root) {
        if (root == null) return null;
        if (root.left != null) { 
            if (root.right != null) 
                root.left.next = root.right; 
            else 
                root.left.next = findNext(root); 
        }
        if (root.right != null) { 
            root.right.next = findNext(root);
        }
        
        connect(root.right); 
        connect(root.left);
        return root;
    }

    private Node findNext(Node root) { 
        while (root.next != null) { 
            root = root.next;
            if (root.left != null) return root.left;
            if (root.right != null) return root.right;
        }
        return null;
    }
}

File: 0117_Populating Next Right Pointers in Each Node II/PopulatingNextRightPointersinEachNodeII.cs
public class Solution {
    public Node Connect(Node root) {
        if(root == null) return root;
        if(root.left!=null)
        {
            if(root.right!=null)
                root.left.next = root.right;
            else
                root.left.next = FindNext(root);
        }
        
        if(root.right!=null)
        {
            root.right.next = FindNext(root);
        }
        
        
        Connect(root.right);
        Connect(root.left);
        
        return root;
    }
    
    private Node FindNext(Node node)
    {
        var p = node;
        while(p.next!=null)
        {
            p = p.next;
            if(p.left != null) return p.left;
            if(p.right != null) return p.right;
        }
        
        return null;
    }
}